---
title: "what I did"
output:
  html_document:
    df_print: paged
date: "2022-07-27"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_chunk$set(eval = FALSE)
```

## Mapping QC

Running fastqc and multiqc

```{bash}
fastqc -t 64 -o ./data/fastqc ./data/raw/*fq.gz
multiqc data/fastqc -o data/multiqc/
```

## Mapping pipeline

Mapping reads to the Mptak_v6.1 genome.

```{bash}
snakemake -c32 -s Snakefile
```

## Post processing of BAM files

```{bash}
#merge bam of HB7
samtools merge HB7.bam HB7*L3.bam HB7*L1.bam
#add RG tag for each bam
bam_array=("HB2" "HB3" "HB5" "HB6" "HB7" "HB8" "HB9")
for i in {0..6}
do
    echo "will add RG to ${bam_array[$i]}.bam"
    bamaddrg -b ${bam_array[$i]}*.bam -s ${bam_array[$i]} -r ${bam_array[$i]} > ${bam_array[$i]}.bam
done
```

## Post analysis of BAM files

### Generating coverage site by site and Displaying distribution in R

```{bash}
samtools depth HB9.bam > HB9.coverage
```

```{r}
coverage_file <- "D:\\LEARN\\project\\generated_data\\RG bam\\HB9.coverage"
coverage=read.table(coverage_file, sep="\t", header=F)

colnames(coverage) <- c(V1="Chr", V2="locus", V3="depth") # renames the header
library(ggplot2)

#ggplot(subset(coverage,Chr=="chr1"), aes(x=locus, y=depth)) +
#geom_point(colour="red", size=1, shape=20, alpha=1/3) 

ggplot(subset(coverage,Chr=="chr1"), aes(x=depth)) +
geom_density(color="cyan2", fill="white") +
xlim(0, 20)

ggplot(subset(coverage,Chr=="chr1"), aes(x=depth)) +
geom_histogram(color="cyan2", fill="white") +
xlim(0, 20)
```

### Plotting coverage along genome by jvarkit

```{bash}
#create a dict of reference genome by picard
java -jar /localdisk/home/s2187039/toolszip/picard.jar CreateSequenceDictionary \
R=/localdisk/home/s2187039/192555/raw/m_project/freebayes_manual_test/RG/MpTak_v6.1.genome.fasta \
O=/localdisk/home/s2187039/192555/raw/m_project/freebayes_manual_test/RG/MpTak_v6.1.genome.dict

#indexing bam
bam_array=("HB2" "HB3" "HB5" "HB6" "HB7" "HB8" "HB9")
for i in {0..6}
do
    echo "will index ${bam_array[$i]}.bam"
    samtools index ${bam_array[$i]}.bam > ${bam_array[$i]}.bam.bai
done

#plot coverage
for i in {0..6}
do
    echo "will plot ${bam_array[$i]}.bam"
    java -jar /localdisk/home/s2187039/toolszip/jvarkit/dist/wgscoverageplotter.jar \
--dimension 1500x500 \
-C -1 \
--clip \
-R /localdisk/home/s2187039/192555/raw/m_project/freebayes_manual_test/RG/MpTak_v6.1.genome.fasta /localdisk/home/s2187039/192555/raw/m_project/freebayes_manual_test/RG/${bam_array[$i]}.bam \
--include-contig-regex "^chr[0-9A-Z]$" --percentile median  > ./bam_coverage_plot/${bam_array[$i]}.bam.svg
done
```

## Variant calling by Freebayes

```{bash}
freebayes -f MpTak_v6.1.genome.fasta -p 1 -v 713.vcf -b *.bam
```

## VCF preliminary filtering

Removing indels and Keeping only known chrs (remove unplaced_scaffolding)

```{bash}
#manual check
bcftools query -f '%CHROM\n' 713*.vcf | uniq #得到那些chrom有snp
bcftools view -r chr1 snp720*.vcf | grep -v -c '^#' #得到chr1 的snp个数

#remove indels
vcftools --vcf 713.vcf --remove-indels --recode --recode-INFO-all --out snp720

#Keep only known chrs (remove unplaced_scaffolding)
vcftools --vcf snp720_recode.vcf --recode --chr chr1 --chr chr2 --chr chr3 --chr chr4 --chr chr5 --chr chr6 --chr chr7 --chr chr8 --chr chrU --chr chrV --recode-INFO-all --out snp726_only_known_chr

```

## VCF sitatistics visualisation

### visualisation in R

Removing lines starts with "\#" in VCF file

```{bash}
#visual the line number of starting with #CHROM 
grep -n "^#CHROM" snp726_only_known_chr.recode.vcf

# the INFO title also starts with #CHROM, so add title to a new file first
#cut the ^# for the title 
grep "^#CHROM" snp726_only_known_chr.recode.vcf | cut -d "#" -f2 > snp726_only_known_chr_no_header.vcf

#grep in VCF, remove ^# and add it to the new file after the title
egrep -v "^#" snp726_only_known_chr.recode.vcf >> snp726_only_known_chr_no_header.vcf
```

Reading VCF into R (limma is needed for strsplit2)

```{r}
if (!require("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

BiocManager::install("limma")
library(limma)
```

```{R}
vcf_file <- "D:\\LEARN\\project\\generated_data\\vcf\\snp726_only_known_chr_no_header.vcf"
vcf <- read.table(vcf_file, header=TRUE)

#first get the info line from vcf 
info <- vcf$INFO  #class(info) ===== "character"

#split by";"
#Strsplit2 () returns data in matrix, making it easier to use the splited results
info_split <- strsplit2(info, split = ";")

nrow(info_split)  #===3259160
ncol(info_split)  #===41
```

Extracting Depth from INFO column

```{r}
#Define a matrix with same rows&columns filled with NA
info_number <- as.data.frame(matrix(NA, nrow(info_split), ncol(info_split))) 

#then split by "=" [,2]==number [,1]==info_of_the_number
for (i in 1:ncol(info_split))
    {
    #define the colnames from the first row
    colnames(info_number)[i] <- strsplit2(info_split[1,i], split = "=")[,1]
    #set the values
    info_number[,i] <- strsplit2(info_split[,i], split = "=")[,2]
    }

#set col names
#colnames(info_number) <- info_colname

#test
info_number$DP[1]
colnames(info_number)
```

#### Plotting DP density

```{r}
#set values
DP <- as.numeric(info_number$DP)
DP <- as.data.frame(DP)
colnames(DP) <- "Depth"

#plot DP density
library("ggplot2")

plot_dp <- ggplot() +
geom_density(data = DP, 
             aes(x = Depth),
             size = 1, 
             color = "black", 
             fill = "darkseagreen1") +
xlim(0, 200) +
geom_segment(aes(mean(DP$Depth), y = 0 ),
           color= "blue",
           size = 1.5,
           xend = mean(DP$Depth),
           yend = 0.035
           ) +
geom_segment(aes(median(DP$Depth), y = 0 ),
           color= "darkorange",
           size = 1.5,
           xend = median(DP$Depth),
           yend = 0.035
           ) +
theme_classic() +
labs(x="Depth", y="Density", title="Depth density", color="") +
  theme(legend.position="none",
        plot.title = element_text(hjust = 0.5),
        axis.title = element_text(size = 15),
        axis.text = element_text(size = 15),
        axis.text.x = element_text(angle = 60, hjust = 1),
        axis.line = element_line(size = 1),
        axis.ticks = element_line(size = 1),
        title = element_text(size = 18))
```

Calculation on Depth

```{r}
#calculate DP
mean(DP$Depth)
median(DP$Depth)
max(DP$Depth)
min(DP$Depth)

nrow(as.data.frame(DP[which(DP > 70),]))
nrow(as.data.frame(DP[which(DP > 70),]))/nrow(DP)

nrow(as.data.frame(DP[which(DP < 5),]))
nrow(as.data.frame(DP[which(DP < 5),]))/nrow(DP)
```

#### Plotting Quality score

```{r}
#set values
QUAL <- as.numeric(vcf$QUAL)
QUAL <- as.data.frame(QUAL)
colnames(QUAL) <- "QUAL"

#plot QUAL density
plot_qual <- ggplot() +
geom_density(data = QUAL, 
             aes(x = QUAL),
             size = 1, 
             color = "black", 
             fill = "darkseagreen1") +
xlim(0, 250) +
geom_segment(aes(mean(QUAL$QUAL), y = 0 ),
           color= "blue",
           size = 1.5,
           xend = mean(QUAL$QUAL),
           yend = 0.065
           ) + 
geom_segment(aes(median(QUAL$QUAL), y = 0 ),
           color= "darkorange",
           size = 1.5,
           xend = median(QUAL$QUAL),
           yend = 0.065
           ) +
theme_classic() +
labs(x="QUAL", y="Density", title="QUAL density") +
  theme(legend.position="top",
        plot.title = element_text(hjust = 0.5),
        axis.title = element_text(size = 15),
        axis.text = element_text(size = 15),
        axis.text.x = element_text(angle = 60, hjust = 1),
        axis.line = element_line(size = 1),
        axis.ticks = element_line(size = 1),
        title = element_text(size = 18))
```

```{r}
library(ggpubr)

png(file= "VCF_DP_QUAL.png")
print(ggarrange(plot_qual, plot_dp,
                labels = c("A","B"),
                nrow = 2,
                ncol = 1))
dev.off()
```

#### plotting QUAL/DP

```{r}
#set values
QUAL_DP <- QUAL/DP
colnames(QUAL_DP) <- "QD"

#plot DP density
ggplot(QUAL_DP, aes(x = QD)) +
geom_density(size = 1) +
geom_density(color = "black", fill = "darkseagreen1") +
xlim(0,40) +
theme_classic() +
labs(x="QUAL devided by DP", y="Density", title="QUAL/DP density", color="") +
  theme(legend.position="none",
        axis.title = element_text(size = 15),
        axis.text = element_text(size = 15),
        axis.text.x = element_text(angle = 60, hjust = 1),
        axis.line = element_line(size = 1),
        axis.ticks = element_line(size = 1),
        title = element_text(size = 12))
```

## VCF hard filtering

Testing thresholds

```{r}
QUAL_threshold <- 30
QUAL_PASS <- as.data.frame(QUAL[which(QUAL >= QUAL_threshold),])

nrow(QUAL_PASS)
nrow(QUAL_PASS)/nrow(QUAL)
nrow(QUAL)

nrow(as.data.frame(QUAL[which(QUAL < 100),]))
```

Filtering by bcftools

```{bash}
bcftools filter -O v -o snp729_bcftools.vcf -i 'QUAL >= 30 && INFO/DP >= 5 && INFO/DP <= 70' snp726*recode.vcf

#check how many SNPs left
egrep -v "^#" snp729_bcftools.vcf | wc -l
#check how many SNPs before filtering for eg chr1
egrep -v "^#" snp726_only_known_chr.recode.vcf | cut -f 1 | grep "chr1" | wc -l
```

Checking statistics of the filtered VCF file

```{bash}
#visual the line number of starting with #CHROM 
grep -n "^#CHROM" snp729_bcftools.vcf

# the INFO title also starts with #CHROM, so add title to a new file first
#cut the ^# for the title 
grep "^#CHROM" snp729_bcftools.vcf | cut -d "#" -f2 > snp729_threshold_DP10to100_Q30_no_header.vcf

#grep in VCF, remove ^# and add it to the new file after the title
egrep -v "^#" snp729_bcftools.vcf >> snp729_bcftools_no_header.vcf
```

```{bash}
# check SNP number for each chromosome
egrep -v "^#" snp729*.vcf | cut -f 1 | grep "chr1" | wc -l
```

```{r}
vcf_file <- "D:\\LEARN\\project\\generated_data\\vcf\\snp729_bcftools_no_header.vcf"
vcf <- read.table(vcf_file, header=TRUE)

#first get the info line from vcf 
info <- vcf$INFO  #class(info) ===== "character"

#split by";"
#Strsplit2 () returns data in matrix, making it easier to use the splited results
info_split <- strsplit2(info, split = ";")

nrow(info_split)  #===1758896
ncol(info_split)  #===41
```

## Calculation of population genetics statistical indicators

### Generating pi by window

A special vcftools is used for haploid data <https://github.com/jydu/vcftools>

```{bash}
vcftools --vcf snp729*.vcf --haploid --window-pi 1000 --out pi1000
#try different window size
vcftools --vcf snp729*.vcf --haploid --window-pi 10000 --out pi10000
```

### Generating Tajima's D by window

```{bash}
vcftools --vcf snp729*.vcf --haploid --TajimaD 1000 --out D_1kb
vcftools --vcf snp729*.vcf --haploid --TajimaD 10000 --out D_10kb
```

### Generating pi by per gene

```{bash}
#generate a bed file first
gff2bed < MpTak_*.gff > gff.bed

#generate pi by per gene
vcftools --haploid --vcf snp729*.vcf --site-pi --bed gff.bed --out pi_per_gene

```

### Generating FST for groups

```{bash}
# group is defined by PCA and heatmap
#group1 hb2 5 6 9 
#group2 hb3,8
#group3 hb7
vcftools --haploid --vcf snp729_bcftools.vcf --weir-fst-pop cluster1.txt --weir-fst-pop cluster2.txt --out fst1-2

vcftools --haploid --vcf snp729_bcftools.vcf --weir-fst-pop cluster1.txt --weir-fst-pop cluster3.txt --out fst1-3

vcftools --haploid --vcf snp729_bcftools.vcf --weir-fst-pop cluster2.txt --weir-fst-pop cluster3.txt --out fst2-3

#random sample 
# sample(c(2,3,5,6,7,8,9),3) ====7 2 6 
vcftools --haploid --vcf snp729_bcftools.vcf --weir-fst-pop random1.txt --weir-fst-pop random2.txt --out fstrandom1-2
# sample(c(2,3,5,6,7,8,9),3) ====2 3 8
vcftools --haploid --vcf snp729_bcftools.vcf --weir-fst-pop random1_1.txt --weir-fst-pop random2_13.txt --out fstrandom3-4
# sample(c(2,3,5,6,7,8,9),3) ====3 7 9
vcftools --haploid --vcf snp729_bcftools.vcf --weir-fst-pop random1_3.txt --weir-fst-pop random2_3.txt --out fstrandom5-6
```

## Detecting selective sweeps by Sweepfinder2

### Generate allele frequency file as the input

read VCF into R and extract genotype data

```{r}
library(vcfR)

vcf_file729 <- "D:\\LEARN\\project\\generated_data\\vcf\\snp729_bcftools.vcf"
vcf729 <- read.vcfR(vcf_file729)
# extract genotype data
gt <- extract.gt(vcf729, element="GT", as.numeric = T)

#test
dim(gt)
tail(gt)
head(gt)
```

any allele with more than one alternative will be removed

```{r}
# remove multi allelic sites
gt[is.na(gt)] <- -1 #set NA to -1
multiallelic <- apply(gt, 1, function(x) sum(x > 1) > 0)#apply by rows

#print statistics
table(multiallelic)
print(paste("the number of allele with more than one alt: ",length(multiallelic[which(multiallelic == TRUE)])))

#any allele have more than one alternative will be removed
gt <- gt[!multiallelic, ]
```

remove chrU and chrV

```{r}
#split chr and chrom position
library(limma)
gt_split <- strsplit2(row.names(gt), split = "_")
gt_split <- as.data.frame(gt_split)
colnames(gt_split) <- c("chrom", "position")

# add into a new table
gt_autochrom <- cbind(gt_split,gt)

#remove chrU and chrV
autosome_number <- gt_autochrom$chrom %in% paste0("chr", 1:8)
gt_autochrom <- gt_autochrom[autosome_number,]
```

generate file with the format that SweepFinder2 requires, which contains four columns, including

position in the genome, the allele count, the sample size, polarized site (1: ancestral allele 0: non-ancestral derived allele).

```{r}
#x == the allele count
#n == the sample size
gt_autochrom$x <- rowSums(gt_autochrom[,3:9] == 1)
gt_autochrom$n <- rowSums(gt_autochrom[,3:9] != -1)
gt_autochrom$folded <- 1
```

save files

```{r}
write.table(gt_autochrom[,c(2, 10, 11, 12)],
            "allChroms",
            row.names = F,
            quote = F,
            sep="\t")

for(i in 1:8){
  chr_name <- paste0("chr", i)
  write.table(gt_autochrom[gt_autochrom$chrom == chr_name,c(2, 10, 11, 12)],
              chr_name,
              row.names = F,
              quote = F,
              sep="\t")
  remove(chr_name)
}
```

### Using sweepfinder2

```{bash}
#first generate empirical frequency spectrum
echo "Computing SFS"
SweepFinder2 -f allChroms spectrum

#analyse for each CHROM
for i in {1..8}
do
  echo "Analysing chromosome $i..."
  SweepFinder2 -l 1000 chr$i spectrum chr$i.out > /dev/null
done
```

## Visualisation of population genetics statistical indicators

### PCA by plink

```{bash}
plink --vcf snp729*.vcf --make-bed --pca --allow-extra-chr --out pca823
plink --vcf snp729*.vcf --make-bed --pca --not-chr chrU,chrV --allow-extra-chr --out pca823_autosome
```

PCA plot for all chromosome result <https://speciationgenomics.github.io/pca/>

```{r}
library(tidyverse)
# read in data
pca <- read_table("pca823.eigenvec", col_names = FALSE)
eigenval <- scan("pca823.eigenval")
# sort out the pca data
# remove nuisance column
pca <- pca[,-1]
# set names
names(pca)[1] <- "ind"
names(pca)[2:ncol(pca)] <- paste0("PC", 1:(ncol(pca)-1))
pca <- as.data.frame(pca)
row.names(pca) <- pca$ind

# first convert to percentage variance explained
pve <- data.frame(PC = 1:7, pve = eigenval/sum(eigenval)*100)


plot_pca1 <- ggplot() +
geom_point(data = pca, 
          aes(x = PC1, y = PC2, color = row.names(pca)), 
          size = 4
          ) +
theme_classic() +
guides(color=guide_legend(title="Samples")) +
labs(title="PCA plot for all chromosomes") +
xlab(paste0("PC1 (", signif(pve$pve[1], 3), "%)")) +
ylab(paste0("PC2 (", signif(pve$pve[2], 3), "%)")) +
  theme(legend.position="top",
        axis.title = element_text(size = 15),
        axis.text = element_text(size = 15),
        axis.line = element_line(size = 1),
        axis.ticks = element_line(size = 1),
        plot.title = element_text(size = 16, hjust=0.5))

plot_pc <- ggplot(pve, aes(PC, pve)) + 
geom_bar(stat = "identity") +
ylab("Variance% explained") +
labs(title="PCA Variance for all chromosomes") +
theme_classic() +
  theme(legend.position="right",
        axis.title = element_text(size = 15),
        axis.text = element_text(size = 15),
        axis.line = element_line(size = 1),
        axis.ticks = element_line(size = 1),
        plot.title = element_text(size = 16, hjust=0.5))
```

PCA plot for only autosomes result

```{r}
# read in data
pca2 <- read_table("pca823_autosome.eigenvec", col_names = FALSE)
eigenval2 <- scan("pca823_autosome.eigenval")
# sort out the pca data
# remove nuisance column
pca2 <- pca2[,-1]
# set names
names(pca2)[1] <- "ind"
names(pca2)[2:ncol(pca2)] <- paste0("PC", 1:(ncol(pca2)-1))
pca2 <- as.data.frame(pca2)
row.names(pca2) <- pca2$ind

# first convert to percentage variance explained
pve2 <- data.frame(PC = 1:7, pve = eigenval2/sum(eigenval2)*100)


plot_pca2 <- ggplot() +
geom_point(data = pca2, 
          aes(x = PC1, y = PC2, color = row.names(pca2)), 
          size = 4
          ) +
theme_classic() +
guides(color=guide_legend(title="Samples")) +
labs(title="PCA plot for autosomes") +
xlab(paste0("PC1 (", signif(pve$pve[1], 3), "%)")) +
ylab(paste0("PC2 (", signif(pve$pve[2], 3), "%)")) +
  theme(legend.position="top",
        axis.title = element_text(size = 15),
        axis.text = element_text(size = 15),
        axis.line = element_line(size = 1),
        axis.ticks = element_line(size = 1),
        plot.title = element_text(size = 16, hjust=0.5))

plot_pc2 <- ggplot(pve2, aes(PC, pve)) + 
geom_bar(stat = "identity") +
ylab("Variance% explained") +
labs(title="PCA Variance for autosomes") +
theme_classic() +
  theme(legend.position="right",
        axis.title = element_text(size = 15),
        axis.text = element_text(size = 15),
        axis.line = element_line(size = 1),
        axis.ticks = element_line(size = 1),
        plot.title = element_text(size = 16, hjust=0.5))
```

```{r}
library(ggpubr)
ggarrange(plot_pca1,plot_pc,plot_pca2,plot_pc2,
            labels = c("A","B","C","D"),
            align = "v",
            nrow = 2,
            ncol = 2)
```

### Heatmap from genotype extracted from VCF

```{r}HB}
heatmap_auto <- pheatmap(na.omit(gt_autochrom[sample(c(1:nrow(gt_autochrom)),size=5000),3:9]),
         cluster_rows = FALSE,
         main = "Heatmap of samples' genotype (no sex chr)", 
         show_rownames = FALSE,
         fontsize = 15,
         treeheight_col = 40)

heatmap_all <- pheatmap(na.omit(gt[sample(c(1:nrow(gt)),size=5000),]),
         cluster_rows = FALSE,
         main = "Heatmap of samples' genotype (all chr)", 
         show_rownames = FALSE,
         fontsize = 15,
         treeheight_col = 40)
```

### Test fst robust

```{r}
#read pi files
fst1_2file <- "D:\\LEARN\\project\\generated_data\\729results\\fst1-2.weir.fst"
fst1_3file <- "D:\\LEARN\\project\\generated_data\\729results\\fst1-3.weir.fst"
fst2_3file <- "D:\\LEARN\\project\\generated_data\\729results\\fst2-3.weir.fst"
fstrandom1file <- "D:\\LEARN\\project\\generated_data\\729results\\fstrandom1-2.weir.fst"
fstrandom2file <- "D:\\LEARN\\project\\generated_data\\729results\\fstrandom3-4.weir.fst"
fstrandom3file <- "D:\\LEARN\\project\\generated_data\\729results\\fstrandom5-6.weir.fst"

fst1_2 <- read.csv(fst1_2file, header = T, sep = "\t")
fst1_3 <- read.csv(fst1_3file, header = T, sep = "\t")
fst2_3 <- read.csv(fst2_3file, header = T, sep = "\t")
fstR1 <- read.csv(fstrandom1file, header = T, sep = "\t")
fstR2 <- read.csv(fstrandom2file, header = T, sep = "\t")
fstR3 <- read.csv(fstrandom3file, header = T, sep = "\t")


sumfst <- as.data.frame(matrix(ncol = 7, nrow = 12))
colnames(sumfst) <- c("CHROM",
                      "fst1_2mean",
                      "fst1_3mean",
                      "fst2_3mean",
                      "fst1_2median",
                      "fst1_3median",
                      "fst2_3median")
sumfst$CHROM <- c(unique(fst1_2$CHROM),"sum_auto","sum_all")

for (chr in unique(fst1_2$CHROM)){
  print(chr)
  sumfst$fst1_2mean[sumfst$CHROM == chr] <- mean(na.omit(fst1_2$WEIR_AND_COCKERHAM_FST[fst1_2$CHROM == chr]))
  sumfst$fst1_3mean[sumfst$CHROM == chr] <- mean(na.omit(fst1_3$WEIR_AND_COCKERHAM_FST[fst1_3$CHROM == chr]))
  sumfst$fst2_3mean[sumfst$CHROM == chr] <- mean(na.omit(fst2_3$WEIR_AND_COCKERHAM_FST[fst2_3$CHROM == chr]))
  
  sumfst$fst1_2median[sumfst$CHROM == chr] <- median(na.omit(fst1_2$WEIR_AND_COCKERHAM_FST[fst1_2$CHROM == chr]))
  sumfst$fst1_3median[sumfst$CHROM == chr] <- median(na.omit(fst1_3$WEIR_AND_COCKERHAM_FST[fst1_3$CHROM == chr]))
  sumfst$fst2_3median[sumfst$CHROM == chr] <- median(na.omit(fst2_3$WEIR_AND_COCKERHAM_FST[fst2_3$CHROM == chr]))  
}

sumfst$fst1_2mean[11] <- mean(na.omit(fst1_2$WEIR_AND_COCKERHAM_FST[!fst1_2$CHROM %in% c("chrU","chrV")]))
sumfst$fst1_3mean[11] <- mean(na.omit(fst1_3$WEIR_AND_COCKERHAM_FST[!fst1_3$CHROM %in% c("chrU","chrV")]))
sumfst$fst2_3mean[11] <- mean(na.omit(fst2_3$WEIR_AND_COCKERHAM_FST[!fst2_3$CHROM %in% c("chrU","chrV")]))
sumfst$fst1_2median[11] <- median(na.omit(fst1_2$WEIR_AND_COCKERHAM_FST[!fst1_2$CHROM %in% c("chrU","chrV")]))
sumfst$fst1_3median[11] <- median(na.omit(fst1_3$WEIR_AND_COCKERHAM_FST[!fst1_3$CHROM %in% c("chrU","chrV")]))
sumfst$fst2_3median[11] <- median(na.omit(fst2_3$WEIR_AND_COCKERHAM_FST[!fst2_3$CHROM %in% c("chrU","chrV")]))

sumfst$fst1_2mean[12] <- mean(na.omit(fst1_2$WEIR_AND_COCKERHAM_FST))
sumfst$fst1_3mean[12] <- mean(na.omit(fst1_3$WEIR_AND_COCKERHAM_FST))
sumfst$fst2_3mean[12] <- mean(na.omit(fst2_3$WEIR_AND_COCKERHAM_FST))
sumfst$fst1_2median[12] <- median(na.omit(fst1_2$WEIR_AND_COCKERHAM_FST))
sumfst$fst1_3median[12] <- median(na.omit(fst1_3$WEIR_AND_COCKERHAM_FST))
sumfst$fst2_3median[12] <- median(na.omit(fst2_3$WEIR_AND_COCKERHAM_FST))


sumfstR <- as.data.frame(matrix(ncol = 7, nrow = 12))
colnames(sumfstR) <- c("CHROM",
                      "fstR1mean",
                      "fstR2mean",
                      "fstR3mean",
                      "fstR1median",
                      "fstR2median",
                      "fstR3median")
sumfstR$CHROM <- c(unique(fstR1$CHROM),"sum_auto","sum_all")

for (chr in unique(fstR1$CHROM)){
  print(chr)
  sumfstR$fstR1mean[sumfstR$CHROM == chr] <- mean(na.omit(fstR1$WEIR_AND_COCKERHAM_FST[fstR1$CHROM == chr]))
  sumfstR$fstR2mean[sumfstR$CHROM == chr] <- mean(na.omit(fstR2$WEIR_AND_COCKERHAM_FST[fstR2$CHROM == chr]))
  sumfstR$fstR3mean[sumfstR$CHROM == chr] <- mean(na.omit(fstR3$WEIR_AND_COCKERHAM_FST[fstR3$CHROM == chr]))
  
  sumfstR$fstR1median[sumfstR$CHROM == chr] <- median(na.omit(fstR1$WEIR_AND_COCKERHAM_FST[fstR1$CHROM == chr]))
  sumfstR$fstR2median[sumfstR$CHROM == chr] <- median(na.omit(fstR2$WEIR_AND_COCKERHAM_FST[fstR2$CHROM == chr]))
  sumfstR$fstR3median[sumfstR$CHROM == chr] <- median(na.omit(fstR3$WEIR_AND_COCKERHAM_FST[fstR3$CHROM == chr]))  
}

sumfstR$fstR1mean[11] <- mean(na.omit(fstR1$WEIR_AND_COCKERHAM_FST[!fstR1$CHROM %in% c("chrU","chrV")]))
sumfstR$fstR2mean[11] <- mean(na.omit(fstR2$WEIR_AND_COCKERHAM_FST[!fstR2$CHROM %in% c("chrU","chrV")]))
sumfstR$fstR3mean[11] <- mean(na.omit(fstR3$WEIR_AND_COCKERHAM_FST[!fstR3$CHROM %in% c("chrU","chrV")]))
sumfstR$fstR1median[11] <- median(na.omit(fstR1$WEIR_AND_COCKERHAM_FST[!fstR1$CHROM %in% c("chrU","chrV")]))
sumfstR$fstR2median[11] <- median(na.omit(fstR2$WEIR_AND_COCKERHAM_FST[!fstR2$CHROM %in% c("chrU","chrV")]))
sumfstR$fstR3median[11] <- median(na.omit(fstR3$WEIR_AND_COCKERHAM_FST[!fstR3$CHROM %in% c("chrU","chrV")]))

sumfstR$fstR1mean[12] <- mean(na.omit(fstR1$WEIR_AND_COCKERHAM_FST))
sumfstR$fstR2mean[12] <- mean(na.omit(fstR2$WEIR_AND_COCKERHAM_FST))
sumfstR$fstR3mean[12] <- mean(na.omit(fstR3$WEIR_AND_COCKERHAM_FST))
sumfstR$fstR1median[12] <- median(na.omit(fstR1$WEIR_AND_COCKERHAM_FST))
sumfstR$fstR2median[12] <- median(na.omit(fstR2$WEIR_AND_COCKERHAM_FST))
sumfstR$fstR3median[12] <- median(na.omit(fstR3$WEIR_AND_COCKERHAM_FST))



write.csv(sumfst,file = "sumfst.csv")
write.csv(sumfstR,file = "sumfstRandom.csv")
```

### read files for pi D and sweep

```{r}
#read pi files
pi1000_file <- "D:\\LEARN\\project\\generated_data\\729results\\pi1000.windowed.pi"
pi10000_file <- "D:\\LEARN\\project\\generated_data\\729results\\pi10000.windowed.pi"
pisite_file <- "D:\\LEARN\\project\\generated_data\\729results\\pi_site_gff.sites.pi"

pi1000 <- read.csv(pi1000_file, header = T, sep = "\t")
pi10000 <- read.csv(pi10000_file, header = T, sep = "\t")
pisite <- read.csv(pisite_file, header = T, sep = "\t")
allwindow_pi <- rbind(transform(pi1000, WINsize = "1000"), transform(pi10000, WINsize = "10000"))


#read D files
d1000_file <- "D:\\LEARN\\project\\generated_data\\729results\\D_1kb.Tajima.D"
d10000_file <- "D:\\LEARN\\project\\generated_data\\729results\\D_10kb.Tajima.D"
d1000 <- read.csv(d1000_file, header=T, sep="\t")
d10000 <- read.csv(d10000_file, header=T, sep="\t")
allwindow_D <- rbind(transform(d1000, WINsize = "1000"), transform(d10000, WINsize = "10000"))
```

```{r}
#sweepfinder results
sweepfinder_path <- "D:\\LEARN\\project\\generated_data\\sweepfinder2\\"
#use eval and parse to read all file in a loop
for (i in 1:8){
  file <- paste(sweepfinder_path,"chr",i,".out",sep = "")
  eval(
    parse(
      text = paste(
        "sweepchr_", i," <- ", 'read.csv(file, header=TRUE, sep="\t")',sep = ""
      )
    )
  )
}

#add a column to identify chrom number
for (i in 1:8){
  sweepchr <- paste("sweepchr_",i,sep = "")
  chrnum <- paste("chr",i,sep = "")
  eval(
    parse(
      text = paste(
        sweepchr,"$CHROM"," <- ", i, sep = ""
      )
    )
  )
}

sweep_chr_all <- rbind(
  sweepchr_1,
  sweepchr_2,
  sweepchr_3,
  sweepchr_4,
  sweepchr_5,
  sweepchr_6,
  sweepchr_7,
  sweepchr_8
)
remove(
  sweepchr_1,
  sweepchr_2,
  sweepchr_3,
  sweepchr_4,
  sweepchr_5,
  sweepchr_6,
  sweepchr_7,
  sweepchr_8
)
```

### plot number of SNPs

```{r}
chrom_position_file1 <- "D:\\LEARN\\project\\generated_data\\729results\\chrom_position_all.csv"

chrom_position1 <- read.csv(chrom_position_file1, header = T, sep = ",")

pi1000_2 <- read.csv(pi1000_file, header = T, sep = "\t")
for (i in 1:(nrow(chrom_position1)-1))
    {
  chr = chrom_position1[(i+1),1] #eg add length of chr1 to numbers of chr2
  chr_plus <- sum(chrom_position1[1:i,2])
  print(chr_plus)
  pi1000_2[which(pi1000_2$CHROM == chr),2] <- pi1000_2[which(pi1000_2$CHROM == chr),2] + chr_plus
  pi1000_2[which(pi1000_2$CHROM == chr),3] <- pi1000_2[which(pi1000_2$CHROM == chr),3] + chr_plus
}

ggplot_Nsnp <- ggplot() +
geom_bar(data = pi1000_2,
           aes(x = BIN_START, y = na.omit(N_VARIANTS), color = CHROM),
           stat="identity", width=0.5) +
theme_classic() +
guides(color=guide_legend(title="Chromosome"),fill="none") +
labs(x="chromosome position",
     y="Number of SNPs", 
     title="Number of SNPs along the chromosome") +
theme(legend.position="top",
        axis.title = element_text(size = 15),
        axis.text = element_text(size = 15),
        axis.text.x = element_blank(),
        axis.text.y = element_text(size = 8),
        axis.line = element_line(size = 1),
        axis.ticks = element_line(size = 1),
        plot.title = element_text(size = 16, hjust=0.5))

ggplot_Nsnp

remove(pi1000_2, chrom_position1, chrom_position_file1)
```

### pre analysis on data

```{r}
#remove NAs
d10000 <-  na.omit(d10000)
d1000 <-  na.omit(d1000)

#remove chrU and chrV
d10000 <- d10000[!(d10000$CHROM %in% c("chrU", "chrV")),]
d1000 <- d1000[!(d1000$CHROM %in% c("chrU", "chrV")),]
pi10000 <- pi10000[!(pi10000$CHROM %in% c("chrU", "chrV")),]
pi1000 <- pi1000[!(pi1000$CHROM %in% c("chrU", "chrV")),]
```

### Sum nSNPs / average Pi and D for each autosome 

```{r}
#set a table to store the summary
snp_sum <- as.data.frame( matrix(nrow = 8) )
rownames(snp_sum) <- unique(pi10000$CHROM)

for (chr in unique(pi10000$CHROM) ){
  print(chr)
  subsnp <- subset(pi10000, CHROM == chr)
  subD <- subset(d10000, CHROM == chr)
  snp_sum[chr,1] <- mean(na.omit(subsnp$PI))
  snp_sum[chr,2] <- mean(na.omit(subD$TajimaD))
  snp_sum[chr,3] <- mean(na.omit(subsnp$N_VARIANTS))/10000
  snp_sum[chr,4] <- sum(na.omit(subsnp$N_VARIANTS))
}


#sum forr all
snp_sum2 <- as.data.frame( matrix() )
rownames(snp_sum2) <- "all"
snp_sum2[1,1] <- mean(na.omit(pi10000$PI))
snp_sum2[1,2] <- mean(na.omit(d10000$TajimaD))
snp_sum2[1,3] <- mean(na.omit(pi10000$N_VARIANTS))/10000
snp_sum2[1,4] <- sum(na.omit(pi10000$N_VARIANTS))

snp_sum <- rbind(snp_sum, snp_sum2)
colnames(snp_sum) <- c("average_PI","average_D", "SNP_persite", "sum_nSNP")
remove(subsnp,chr,snp_sum2)
write.csv(snp_sum,file = "Sum_of_Nsnp.csv")
```

### loess values fo pi10000 and d10000 and sweepfinder

```{r}
#pi
#set a new column to store smooth values
smooth_pi10000 <- data.frame(matrix(NA,nrow(pi10000),1))
colnames(smooth_pi10000) <- "Smoothed"
pi10000 <- cbind(pi10000,smooth_pi10000)
remove(smooth_pi10000)

#smooth data for chr1 to 8
for (i in 1:8){
  smooth_value <- predict(
    loess(
      PI ~ BIN_START,
      data = subset(pi10000, CHROM == paste0("chr",i)),
      span=0.08
    )
  )
  pi10000[which(pi10000$CHROM == paste0("chr",i)),6] <- smooth_value
  remove(smooth_value)
}


#loess TajimaD
#set a new column to store smooth values
smooth_d10000 <- data.frame(matrix(NA,nrow(d10000),1))
colnames(smooth_d10000) <- "Smoothed"
d10000 <- cbind(d10000,smooth_d10000)
remove(smooth_d10000)

#smooth data for chr1 to 8
for (i in 1:8){
  s <- predict(
    loess(
      TajimaD ~ BIN_START,
      data = subset(d10000, CHROM == paste0("chr",i)),
      span=0.08
    )
  )
  d10000$Smoothed[which(d10000$CHROM == paste0("chr",i))] <- s
}


#loess sweepfinder
#set a new column to store smooth values
smooth_sweep <- data.frame(matrix(NA,nrow(sweep_chr_all),1))
colnames(smooth_sweep) <- "Smoothed"
sweep_chr_all <- cbind(sweep_chr_all,smooth_sweep)
remove(smooth_sweep)

#smooth data for chr1 to 8
for (i in 1:8){
  s <- predict(
    loess(
      LR ~ location,
      data = subset(sweep_chr_all, CHROM == i),
      span=0.14
    )
  )
  sweep_chr_all$Smoothed[which(sweep_chr_all$CHROM == i)] <- s
}

#if smooth LR < 0 set to 0
sweep_chr_all$Smoothed[sweep_chr_all$Smoothed < 0] <- 0

#add a color columns
#color1 for chr
for (i in 1:8){
  sweep_chr_all$color[sweep_chr_all$CHROM == i] <- paste("chr",i,sep = "")
}

#color2 for threshold
sweep_chr_all$color2 <- sweep_chr_all$color
sweep_chr_all$color2[which(sweep_chr_all$LR > 20)] <- "red" 
```

### judge TajimaD \>0 or \<0

```{r}
d10000$judge <- d10000$CHROM
d10000$judge[which(d10000$TajimaD > 0)] <- "cyan2"
d10000$judge[which(d10000$TajimaD < 0)] <- "chartreuse"
```

### plot density

```{r}
library(ggplot2)
# plot all window size pi density
pi_density <- ggplot() +
geom_density(data = allwindow_pi, aes(x = PI, color = WINsize), size = 1.3, show.legend= TRUE) +
theme_classic() +
guides(color=guide_legend(title="Window size")) +
labs(x="PI", y="Density", title="Distribution of PI values") +
  theme(legend.position=c(.8, .9),
        axis.title = element_text(size = 15),
        axis.text = element_text(size = 15),
        axis.text.x = element_text(angle = 60, hjust = 1),
        axis.line = element_line(size = 1),
        axis.ticks = element_line(size = 1),
        plot.title = element_text(size = 16, hjust=0.5))

#plot Tajima's D density
d_density <- ggplot() +
geom_density(data = allwindow_D, aes(x = TajimaD, color = WINsize), size = 1.3, show.legend= TRUE) +
theme_classic() +
guides(color=guide_legend(title="Window size")) +
labs(x="Tajima's D", y="Density", title="Distribution of Tajima's D") +
  theme(legend.position=c(.8, .9),
        axis.title = element_text(size = 15),
        axis.text = element_text(size = 15),
        axis.text.x = element_text(angle = 60, hjust = 1),
        axis.line = element_line(size = 1),
        axis.ticks = element_line(size = 1),
        plot.title = element_text(size = 16, hjust=0.5))



ggarrange(pi_density, d_density,
          labels = c("A","B"),
          nrow = 2,
          align = "v",
          ncol = 1)
```

### Arranging pi and D , sweep positions for autosomes

read chromosome length

```{r}
chrom_position_file <- "D:\\LEARN\\project\\generated_data\\729results\\chrom_position.csv"

chrom_position <- read.csv(chrom_position_file, header = T, sep = ",")
chrom_position
```

make all chromosome arranged on the x-axis from smallest to largest

```{r}
#add previous chromosome length to chromosome 
#eg for chr8 add (sum(length(chr1:7)))

for (i in 1:(nrow(chrom_position)-1))
    {
  chr = chrom_position[(i+1),1] #eg add length of chr1 to numbers of chr2
  chr_plus <- sum(chrom_position[1:i,2])
  print(chr_plus)
  pi10000[which(pi10000$CHROM == chr),2] <- pi10000[which(pi10000$CHROM == chr),2] + chr_plus
  pi10000[which(pi10000$CHROM == chr),3] <- pi10000[which(pi10000$CHROM == chr),3] + chr_plus
}

for (i in 1:(nrow(chrom_position)-1))
    {
  chr = chrom_position[(i+1),1] #eg add length of chr1 to numbers of chr2
  chr_plus <- sum(chrom_position[1:i,2])
  print(chr_plus)
  d10000[which(d10000$CHROM == chr),2] <- d10000[which(d10000$CHROM == chr),2] + chr_plus
}


#arrange chrom position from chr1 to chr8
for (i in 1:(nrow(chrom_position)-1))
    {
  chr = i+1 #eg add length of chr1 to numbers of chr2
  chr_plus <- sum(chrom_position[1:i,2]) #length of sum of previous chr length 
  print(chr_plus)
  sweep_chr_all[which(sweep_chr_all$CHROM == chr),1] <- sweep_chr_all[which(sweep_chr_all$CHROM == chr),1] + chr_plus
}
```

### Plotting along the chromosome

```{r}
#pi plot
plot_pi_allchr <-ggplot() +
geom_path(data = pi10000, 
          aes(x = BIN_START, y = PI, color = CHROM), 
          size = 1) +
geom_line(data = pi10000, 
          aes(x = BIN_START, y = Smoothed), 
          size=1, 
          color= "black") +
geom_segment(aes(x = pi10000$BIN_START[1], y = mean(pi10000$PI) ),
           color= "blue",
           size = 2,
           xend = tail(pi10000$BIN_START,n=1),
           yend = mean(pi10000$PI)
           ) +
theme_classic() +
guides(color=guide_legend(title="Chromosome")) +
labs(x="chromosome position", y="PI", title="PI plot of all chromosomes (10kb)") +
  theme(legend.position="top",
        axis.title = element_text(size = 15),
        axis.text = element_text(size = 15),
        axis.text.x = element_blank(),
        axis.line = element_line(size = 1),
        axis.ticks = element_line(size = 1),
        plot.title = element_text(size = 16, hjust=0.5))



#TajimaD plot
plot_D_allchr <- ggplot() +
geom_path(size = 1,
          data = d10000, 
          aes(x = BIN_START, y = TajimaD, color = CHROM)
          ) +
geom_line(data = d10000, 
          aes(x = BIN_START, y = Smoothed), 
          size=1, 
          color= "black") +
geom_segment(aes(x = d10000$BIN_START[1], y = mean(d10000$TajimaD) ),
           color= "blue",
           size = 2,
           xend = tail(d10000$BIN_START,n=1),
           yend = mean(d10000$TajimaD)
           ) +
theme_classic() +
guides(color=guide_legend(title="Chromosome")) +
labs(x="chromosome position", y="Tajima's D", title="Tajima's D plot of all chromosomes (10kb)") +
  theme(legend.position="top",
        axis.title = element_text(size = 15),
        axis.text = element_text(size = 15),
        axis.text.x = element_blank(),
        axis.line = element_line(size = 1),
        axis.ticks = element_line(size = 1),
        plot.title = element_text(size = 16, hjust=0.5))



#sweep plot
plot_sweep_allchr <- ggplot() +
geom_point(data = sweep_chr_all,
           aes(x = location, y = LR, color = color),
           size = 1) +
geom_line(data = sweep_chr_all,
          aes(x = location, y = alpha),
          size = 0.9,
          color = "blue") +
geom_line(data = sweep_chr_all, 
          aes(x = location, y = Smoothed), 
          size=0.8, 
          color= "black") +
ylim(0,600)+
theme_classic() +
guides(color=guide_legend(title="Chromosome")) +
labs(x="chromosome position", y="Likelihood ratio", title="Sweepfinder2 LR of all chromosomes") +
  theme(legend.position="top",
        axis.title = element_text(size = 15),
        axis.text = element_text(size = 15),
        axis.text.x = element_blank(),
        axis.text.y = element_text(size = 13),
        axis.line = element_line(size = 1),
        axis.ticks = element_line(size = 1),
        plot.title = element_text(size = 16, hjust=0.5))


plot_sweep_allchr2 <- ggplot() +
geom_point(data = sweep_chr_all,
           aes(x = location, y = log2(LR), color = color),
           size = 1) +
geom_line(data = sweep_chr_all,
          aes(x = location, y = log2(alpha)),
          size = 0.9,
          color = "blue",
          fill = " blue") +
geom_line(data = sweep_chr_all, 
          aes(x = location, y = log2(Smoothed)), 
          size=1.1, 
          color= "black") +
ylim(0,10)+
theme_classic() +
guides(color=guide_legend(title="Chromosome")) +
labs(x="chromosome position", y="LR in Log2", title="Log Sweepfinder2 LR of all chromosomes") +
  theme(legend.position="top",
        axis.title = element_text(size = 15),
        axis.text = element_text(size = 15),
        axis.text.x = element_blank(),
        axis.text.y = element_text(size = 13),
        axis.line = element_line(size = 1),
        axis.ticks = element_line(size = 1),
        plot.title = element_text(size = 16, hjust=0.5))



library(ggpubr)
ggarrange(plot_pi_allchr, plot_D_allchr,
          labels = c("A","B"),
          align = "v",
          nrow = 2,
          ncol = 1
)

ggarrange(plot_sweep_allchr, plot_sweep_allchr2,
          labels = c("A","B"),
          nrow = 2,
          align = "v",
          ncol = 1
)
```

### Finding interested regions of [smoothed LR \> threshold(20)]

```{r}

# threshold
sweep_threshold <- 20
sweep_chr_all$region <- 0

#extract regions by smoothed
for (i in 1:nrow(sweep_chr_all)){
  if (isTRUE((sweep_chr_all$Smoothed[i] < sweep_threshold) && 
      (sweep_chr_all$Smoothed[i+1] >= sweep_threshold) &&
      (sweep_chr_all$Smoothed[i+2] >= sweep_threshold)))
    {
    sweep_chr_all$region[i] <- "start"
  }
  if (isTRUE((sweep_chr_all$Smoothed[i] >= sweep_threshold) && 
      (sweep_chr_all$Smoothed[i+1] >= sweep_threshold)))
    {
    sweep_chr_all$region[i] <- "mid"
  }
  if (isTRUE((sweep_chr_all$Smoothed[i+1] <= sweep_threshold) && 
      (sweep_chr_all$Smoothed[i] > sweep_threshold) &&
      (sweep_chr_all$Smoothed[i-1] > sweep_threshold)))
    {
    sweep_chr_all$region[i+1] <- "end"
    sweep_chr_all$region[i] <- "mid"
  }  
}

```

```{r}
# threshold
sweep_threshold <- 20
sweep_chr_all$region2 <- 0

#extract regions
for (i in 1:nrow(sweep_chr_all)){
  if (isTRUE((sweep_chr_all$LR[i] < sweep_threshold) && 
      (sweep_chr_all$LR[i+1] >= sweep_threshold) &&
      (sweep_chr_all$LR[i+2] >= sweep_threshold)))
    {
    sweep_chr_all$region2[i] <- "start"
  }
  if (isTRUE((sweep_chr_all$LR[i] >= sweep_threshold) && 
      (sweep_chr_all$LR[i+1] >= sweep_threshold)))
    {
    sweep_chr_all$region2[i] <- "mid"
  }
  if (isTRUE((sweep_chr_all$LR[i+1] <= sweep_threshold) && 
      (sweep_chr_all$LR[i] > sweep_threshold) &&
      (sweep_chr_all$LR[i-1] > sweep_threshold)))
    {
    sweep_chr_all$region2[i+1] <- "end"
    sweep_chr_all$region2[i] <- "mid"
  }  
}

```

### Plot along each chrom

```{r}
library(ggpubr)
library(ggplot2)

for (i in 1:8){
  chrnum <- paste0("chr",i)
  
  png(file= paste0("indicator_", chrnum, ".png"),
      width = 400, 
      height = 500)
#pi plot
  plot_pi_each <- ggplot() +
  geom_point(data = subset(pi10000, CHROM == chrnum),
             aes(x = BIN_START, y = PI, color = CHROM),
             size = 1) +
  geom_line(data = subset(pi10000, CHROM == chrnum), 
            aes(x = BIN_START, y = Smoothed), 
            size=1, 
            color= "black") +
  theme_classic() +
  guides(color=guide_legend(title="Chromosome")) +
  labs(x="chromosome position",
       y="PI", 
       title= paste0("PI plot of ", chrnum)) +
  theme(legend.position="None",
        axis.title = element_text(size = 15),
        axis.text = element_text(size = 15),
        axis.text.x = element_blank(),
        axis.text.y = element_text(size = 10),
        axis.line = element_line(size = 1),
        axis.ticks = element_line(size = 1),
        plot.title = element_text(size = 16, hjust=0.5))

#d plot
  plot_d_each <- ggplot() +
  geom_point(data = subset(d10000, CHROM == chrnum),
             aes(x = BIN_START, y = TajimaD, color = judge),
             size = 1) +
  geom_line(data = subset(d10000, CHROM == chrnum), 
            aes(x = BIN_START, y = Smoothed), 
            size=1, 
            color= "black") +
  geom_hline(yintercept = 0, color = "blue") +
  theme_classic() +
  guides(color=guide_legend(title="Chromosome")) +
  labs(x="chromosome position",
       y="Tajima's D", 
       title= paste0("Tajima's D plot of ", chrnum)) +
  theme(legend.position="None",
        axis.title = element_text(size = 15),
        axis.text = element_text(size = 15),
        axis.text.x = element_blank(),
        axis.text.y = element_text(size = 11),
        axis.line = element_line(size = 1),
        axis.ticks = element_line(size = 1),
        plot.title = element_text(size = 16, hjust=0.5))  

#sweep plot
  ylimit <- max(subset(sweep_chr_all, CHROM == i)$Smoothed)+30
  
  plot_sweep_each <- ggplot() +
  geom_bar(data = subset(sweep_chr_all, CHROM == i), 
           aes(x = location, y = Smoothed), 
           stat = 'identity',
           fill = "azure3") +
  geom_bar(data = subset(sweep_chr_all, CHROM == i & region == "mid"), 
            aes(x = location, y = Smoothed), 
            stat = 'identity', 
            color = "darkgoldenrod1") +
  geom_bar(data = subset(sweep_chr_all, CHROM == i & region2 == "mid"), 
            aes(x = location, y = ylimit-30), 
            stat = 'identity', 
            color = "chartreuse3") +
  geom_line(data = subset(sweep_chr_all, CHROM == i),
             aes(x = location, y = Smoothed),
             size = 1,
             color = "black") +  
  geom_point(data = subset(sweep_chr_all, CHROM == i),
             aes(x = location, y = LR, color = color2),
             size = 1) +
  ylim(0,ylimit) +
  theme_classic() +
  guides(color=guide_legend(title="Chromosome")) +
  labs(x="chromosome position",
       y="SweepFinder2 LR", 
       title= paste0("Likelihood ratio plot of ", chrnum)) +
  theme(legend.position="None",
        axis.title = element_text(size = 15),
        axis.text = element_text(size = 15),
        axis.text.x = element_blank(),
        axis.text.y = element_text(size = 11),
        axis.line = element_line(size = 1),
        axis.ticks = element_line(size = 1),
        plot.title = element_text(size = 16, hjust=0.5)) 


  print(ggarrange(plot_sweep_each, plot_d_each, plot_pi_each,
            labels = c("A","B","C"),
            align = "v",
            nrow = 3,
            ncol = 1))
  dev.off()

}


```

## Gathering gene information

### Plot gene information from gff file

```{r}
if (!require("BiocManager", quietly = TRUE))
    install.packages("BiocManager")
BiocManager::install("rtracklayer")
BiocManager::install("karyoploteR")


library(rtracklayer) #for import.gff3(gff_file)
library(ape) #for read.gff()
library(karyoploteR) #for genome plot
#library(AcidGenomes) #for makeGRangesFromGFF()

#read gff file as a GRanges
gff_file <- "D:\\LEARN\\project\\R_workingdir\\MpTak_v6.1r1.gff"
machantia_grange <- import.gff3(gff_file)
machantia_grange_genes <- machantia_grange[machantia_grange$type == "gene"]

#read chromsome length 
#the chrom length was got from the VCF header and saved as a file 
chrom_position_all_file <- "D:\\LEARN\\project\\generated_data\\729results\\chrom_position_all.csv"

chrom_position_all <- read.csv(chrom_position_all_file, header = T, sep = ",")
machantia_chr <- as.data.frame(matrix(nrow = 10,ncol = 3))
colnames(machantia_chr) <- c("chr", "start", "end")
machantia_chr$chr <- chrom_position_all$CHROM
machantia_chr$end <- chrom_position_all$length
machantia_chr$start <- 1

### plot by karyoploteR ###
# the code for karyoploteR is modified from https://bernatgel.github.io/karyoploter_tutorial//Examples/GeneDensityIdeograms/GeneDensityIdeograms.html 

## plot genome with gene position
#set plot.params 
pp <- getDefaultPlotParams(plot.type=2)
pp$data1outmargin <- 100
pp$data2outmargin <- 100
pp$topmargin <- 450

#plot gene positions
kp <- plotKaryotype(genome = machantia_chr, 
                    ideogram.plotter = NULL,
                    plot.type=2,
                    plot.params = pp)
kpAddCytobandsAsLine(kp)
kpAddMainTitle(kp, "Machantia polymorpha genome with genes", cex=1.5)
kpPlotRegions(kp, 
              data = machantia_grange_genes[strand(machantia_grange_genes)=="+"],
              avoid.overlapping = TRUE,
              col="darkgoldenrod3")
kpPlotRegions(kp, 
              data = machantia_grange_genes[strand(machantia_grange_genes)=="-"],
              avoid.overlapping = TRUE,
              data.panel=2,
              col="#3388FF")
kpAddLabels(kp, "strand +", 
            cex=0.8, 
            col="darkgoldenrod1")
kpAddLabels(kp, "strand -", 
            data.panel=2, 
            cex=0.8, 
            col="#3388FF")

#plot gene density
kp <- plotKaryotype(genome = machantia_chr,
                    plot.type=6, 
                    main="Machantia polymorpha Gene Density",
                    ideogram.plotter = NULL,
                    cex=1.8)

kpDataBackground(kp, 
                 color = "#FFFFFFAA")
kp <- kpPlotDensity(kp, 
                    machantia_grange_genes, 
                    window.size = 0.8e5, 
                    data.panel="ideogram", 
                    col="#3388FF", 
                    border="#3388FF", 
                    r0=0.5, r1=1)
kp <- kpPlotDensity(kp, 
                    machantia_grange_genes, 
                    window.size = 0.8e5, 
                    data.panel="ideogram", 
                    col="#3388FF", 
                    border="#3388FF", 
                    r0=0.5, r1=0)

```

### Summarize gene information

```{r}
#read gff as a table
library(ape)
gff_file <- "D:\\LEARN\\project\\R_workingdir\\MpTak_v6.1r1.gff"
gene_summary <- read.gff(gff_file, GFF3 = TRUE)
gene_summary <- gene_summary[gene_summary$type == "gene",]
gene_summary$length <- (gene_summary$end - gene_summary$start)
gene_summary <- gene_summary[gene_summary$seqid %in% unique(chrom_position_all$CHROM),]

#summary
gene_summary_table <- as.data.frame(matrix(nrow = 11,ncol = 9))
rownames(gene_summary_table)[1] <- "All"
rownames(gene_summary_table)[2:11] <- unique(chrom_position_all$CHROM)
colnames(gene_summary_table) <- c("number", "+", "-", "average_length", "max", "min", "total", "chr_length", "gene/contig")


for (i in 1:10){
  chr <- rownames(gene_summary_table)[i+1]
  gene_chr <- gene_summary[gene_summary$seqid == chr,]
  gene_summary_table[i+1,1] <- nrow(gene_chr)
  gene_summary_table[i+1,2] <- nrow(gene_chr[gene_chr$strand == "+",])
  gene_summary_table[i+1,3] <- gene_summary_table[i+1,1] - gene_summary_table[i+1,2]
  gene_summary_table[i+1,4] <- mean(gene_chr$length[gene_chr$seqid == chr])
  gene_summary_table[i+1,5] <- max(gene_chr$length[gene_chr$seqid == chr])
  gene_summary_table[i+1,6] <- min(gene_chr$length[gene_chr$seqid == chr])
  gene_summary_table[i+1,7] <- sum(gene_chr$length[gene_chr$seqid == chr])
}
gene_summary_table$chr_length[2:11] <- chrom_position_all$length



gene_summary_table[1,1] <- sum(gene_summary_table[2:11,1])
gene_summary_table[1,2] <- sum(gene_summary_table[2:11,2])
gene_summary_table[1,3] <- sum(gene_summary_table[2:11,3])
gene_summary_table[1,4] <- mean(gene_summary$length)
gene_summary_table[1,5] <- max(gene_summary_table[2:11,5])
gene_summary_table[1,6] <- min(gene_summary_table[2:11,6])
gene_summary_table[1,7] <- sum(gene_summary_table[2:11,7])
gene_summary_table[1,8] <- sum(gene_summary_table[2:11,8])
gene_summary_table[,9] <- gene_summary_table[,7]/gene_summary_table[,8]

write.csv(file = "gene_summary.csv", gene_summary_table)


#plot gene length density for autosomes
library(ggplot2)
autosomes <- subset(gene_summary, !seqid %in% c("chrU", "chrV"))
plot_auto <- ggplot() +
geom_density(data = autosomes, 
             aes(x = length), 
             size = 1.3) +
geom_segment(aes(x = mean(autosomes$length), y = 0 ),
           color= "blue",
           size = 1.2,
           xend = mean(autosomes$length),
           yend = 0.00022
           ) +
theme_classic() +
labs(x="Gene length", y="Density", title="Autosomes' Gene length density") +
  theme(
        axis.title = element_text(size = 15),
        axis.text = element_text(size = 15),
        axis.text.x = element_text(angle = 60, hjust = 1),
        axis.line = element_line(size = 1),
        axis.ticks = element_line(size = 1),
        plot.title = element_text(size = 20, hjust=0.5))

#plot gene length density for autosomes
sexsomes <- subset(gene_summary, seqid %in% c("chrU", "chrV"))
plot_sex <- ggplot() +
geom_density(data = sexsomes, 
             aes(x = length), 
             size = 1.3) +
geom_segment(aes(x = mean(sexsomes$length), y = 0 ),
           color= "blue",
           size = 1.2,
           xend = mean(sexsomes$length),
           yend = 0.00022
           ) +
theme_classic() +
labs(x="Gene length", y="Density", title="Sex chromosomes' Gene length density") +
  theme(
        axis.title = element_text(size = 15),
        axis.text = element_text(size = 15),
        axis.text.x = element_text(angle = 60, hjust = 1),
        axis.line = element_line(size = 1),
        axis.ticks = element_line(size = 1),
        plot.title = element_text(size = 20, hjust=0.5))

png(filename = "gene_density.png")
print(ggarrange(plot_auto, plot_sex,
            labels = c("A","B"),
            align = "v",
            nrow = 2,
            ncol = 1))
dev.off()

remove(autosomes)
remove(sexsomes)
```

### Estimate pi for genes and save to genetable

```{r}
gff_file <- "D:\\LEARN\\project\\R_workingdir\\MpTak_v6.1r1.gff"

gff <- read.table(gff_file, sep="\t")

#extraact only gene info
gff2 <- gff[which(gff$V3 == "gene"),]
remove(gff)
rownames(gff2) <- NULL
library(limma)
#extract ID fro column 9
for (i in 1:nrow(gff2)) {
  if (strsplit2(gff2[i,9], split = "=")[1] == "ID"){
    gff2[i,10] <- strsplit2(
                          strsplit2(gff2[i,9], split = "=")[2],
                          split = ";"
                         )[1]
  }
}

#set col names
colnames(gff2) <- c("CHROM", "source", "type", "start", "end", "score", "strand", "phase", "attributes", "ID")

#remove genes on unplaced-scaffold
gff2 <- gff2[-(grep('^unplaced-scaffold',gff2[,1])),]

#add pi and D values
pi1000_file <- "D:\\LEARN\\project\\generated_data\\729results\\pi1000.windowed.pi"
pi_table <- read.csv(pi1000_file, header = T, sep = "\t")

#subset gene and pi by CHROM, set 10 tables to store the value
gene_chr = list()
gff_pi = list()
pi_chr = list()
for (chr in unique(gff2[,1])) {
  gene_chr[[chr]] <- subset(gff2, CHROM == chr)
  pi_chr[[chr]] <- subset(pi_table, CHROM == chr)
  gff_pi[[chr]] <- as.data.frame(matrix(nrow = nrow( gene_chr[[chr]] ), ncol = 7))
  colnames( gff_pi[[chr]] ) <- c("pi_start","pi_end","start_row", "end_row", "BINS", "pi_for_gene", "nSNPs")
  gff_pi[[chr]]$BINS <- 0
}

#calculate pi 
for (chr in unique(gff2[,1])){
  gene <- gene_chr[[chr]]
  pi <- pi_chr[[chr]]
  print(chr)
  
  for (i in 1:nrow(gene)){
    for (j in 1:nrow(pi)){
      # if pi window start < gene start < pi window end
      if (gene[i,4] >= pi[j,2] && gene[i,4] <= pi[j,3]){
        gff_pi[[chr]]$pi_start[i] = pi[j,2]
        gff_pi[[chr]]$start_row[i] = j
      }
      # if no (pi window start < gene start < pi window end) found but
      # gene start < pi window start && pi window end < gene end
      if (gene[i,4] < pi[j,2] && gene[i,5] > pi[j,3] ){
        gff_pi[[chr]]$BINS[i] = gff_pi[[chr]]$BINS[i] + 1

      }
      
      #if pi window start < gene end < pi window end
      if (gene[i,5] >= pi[j,2] && gene[i,5] <= pi[j,3]){
        gff_pi[[chr]]$pi_end[i] = pi[j,3]
        gff_pi[[chr]]$end_row[i] = j
        break
      }
    }
    
    #calculation on pi for each gene
    #FOUR situation
    # 1. no pi window inside --->no calculation
    # 2. gene start is in pi window, but gene end not in pi window
    # 3. no gene start is in pi window, but gene end  in pi window
    # 4. gene start and end both in pi window (in same window or not)
    
    # 4. gene start and end both in pi window (in same window or not)
    if ( (!is.na(gff_pi[[chr]]$start_row[i])) && 
         (!is.na(gff_pi[[chr]]$end_row[i])) ){
      
      gff_pi[[chr]]$pi_for_gene[i] = mean(
        pi$PI[ gff_pi[[chr]]$start_row[i] : gff_pi[[chr]]$end_row[i] ])
      
      gff_pi[[chr]]$nSNPs[i] = sum(
        pi$N_VARIANTS[ gff_pi[[chr]]$start_row[i] : gff_pi[[chr]]$end_row[i] ])
      
    } 
    
    # 2. gene start is in pi window, but gene end not in pi window
    else if ( (!is.na(gff_pi[[chr]]$start_row[i])) && 
              (is.na(gff_pi[[chr]]$end_row[i])) ) {
      
      gff_pi[[chr]]$pi_for_gene[i] = mean(
        pi$PI[ gff_pi[[chr]]$start_row[i] : 
                 (gff_pi[[chr]]$start_row[i] + gff_pi[[chr]]$BINS[i] ) ])
      
      gff_pi[[chr]]$nSNPs[i] = sum(
        pi$N_VARIANTS[ gff_pi[[chr]]$start_row[i] : 
                         (gff_pi[[chr]]$start_row[i] + gff_pi[[chr]]$BINS[i] ) ])
    } 
    
    # 3. no gene start is in pi window, but gene end  in pi window    
    else if ( (!is.na(gff_pi[[chr]]$end_row[i])) && 
              (is.na(gff_pi[[chr]]$start_row[i])) ) {
      
      gff_pi[[chr]]$pi_for_gene[i] = mean(
        pi$PI[ (gff_pi[[chr]]$end_row[i] - gff_pi[[chr]]$BINS[i]) : 
                 gff_pi[[chr]]$end_row[i] ])
      
      gff_pi[[chr]]$nSNPs[i] = sum(
        pi$N_VARIANTS[ (gff_pi[[chr]]$end_row[i] - gff_pi[[chr]]$BINS[i]) : 
                 gff_pi[[chr]]$end_row[i] ])
    }
    
    print(i)
  }
  print(paste0("calculation on ", chr, " is end"))
}



pppp <- rbind(gff_pi[["chr1"]],
              gff_pi[["chr2"]],
              gff_pi[["chr3"]],
              gff_pi[["chr4"]],
              gff_pi[["chr5"]],
              gff_pi[["chr6"]],
              gff_pi[["chr7"]],
              gff_pi[["chr8"]],
              gff_pi[["chrU"]],
              gff_pi[["chrV"]]
              )



#gather data
my_gff <- cbind(gff2, pppp$pi_for_gene, pppp$nSNPs)
colnames(my_gff)[11:12] <- c("PI" , "NUMBERofSNPs")

write.csv(my_gff, file = "gene_table_pi.csv", row.names = F)
```

### Summarizing pi for gene table

```{r}
#from my_gff

#set a table to store the summary
gene_sum <- as.data.frame( matrix(nrow = 10) )
rownames(gene_sum) <- unique(my_gff$CHROM)

for (chr in unique(my_gff$CHROM) ){
  print(chr)
  subgene <- subset(my_gff, CHROM == chr)
  gene_sum[chr,1] <- mean(na.omit(subgene$PI))
  gene_sum[chr,2] <- mean(na.omit(subgene$NUMBERofSNPs))
}


#sum forr all
gene_sum2 <- as.data.frame( matrix() )
rownames(gene_sum2) <- "all"
gene_sum2[1,1] <- mean(na.omit(my_gff$PI))
gene_sum2[1,2] <- mean(na.omit(my_gff$NUMBERofSNPs))

gene_sum <- rbind(gene_sum, gene_sum2)
colnames(gene_sum) <- c("average_PI","average_nSNP")
remove(subgene,chr,gene_sum2)

write.csv(gene_sum,file = "Sum_of_pi_for_Gene_table.csv")
```

## Checking genes in the interested regions

### Take the intersection of smoothed LR regions and LR regions

```{r}
#use sweep_chr_all and gene_summary


library(limma)
gene_summary$ID <- NA
#extract ID as column 9
for (i in 1:nrow(gene_summary)) {
  if (strsplit2(gene_summary$attributes[i], split = "=")[1] == "ID"){
    gene_summary$ID[i] <- strsplit2(
                          strsplit2(gene_summary$attributes[i], split = "=")[2],
                          split = ";"
                         )[1]
  }
}

#taking the intersection of smoothed LR regions and LR regions
regions_smooth <- subset(sweep_chr_all, region == "start" | region == "end" | region == "mid")
region_limit <- subset(regions_smooth, region2 != 0)
```

### Find genes in the region

```{r}
#my file was generated manually from region_limit 
my_region_file <- "D:\\LEARN\\project\\R_workingdir\\myregion2.csv"
my_region <- read.csv(my_region_file,header = T)

#calculate info for regions
for (i in 1:nrow(my_region)){
    my_region$averageLR[i] <- mean(sweep_chr_all$LR[ my_region$index.start[i]:my_region$index.end[i] ])
    my_region$averageAlpha[i] <- mean(sweep_chr_all$alpha[ my_region$index.start[i]:my_region$index.end[i] ])
    my_region$peak.LR[i] <- max(sweep_chr_all$LR[ my_region$index.start[i]:my_region$index.end[i] ])
    my_region$peak.Alpha[i] <- max(sweep_chr_all$alpha[ my_region$index.start[i]:my_region$index.end[i] ])
}



#find genes in the region
gene_in_region <- list()
for (i in 1:nrow(my_region)){
  chr <- my_region$CHROM[i]
  sub_gene <- subset(gene_summary, seqid == chr)
  sub_gene$mid <- (sub_gene$start + sub_gene$end)/2
  gene_in_region[[i]] <- vector()
  for (j in 1:nrow(sub_gene)){
    if (isTRUE(
      ((sub_gene$start[j] > my_region$start[i]) && 
       (sub_gene$start[j] < my_region$end[i])) ||
      ((sub_gene$end[j] > my_region$start[i]) &&
       (sub_gene$end[j] < my_region$end[i])) ||
      ((my_region$start[i] > sub_gene$start[j]) &&
       (my_region$end[i] < sub_gene$end[j]))
    )){
      gene_in_region[[i]] <- append(gene_in_region[[i]], 
                     sub_gene$ID[j],
                     after = length(gene_in_region[[i]]))
      }
  }
}



#find max number of genes among each region
for (i in 1:nrow(my_region)){
  my_region$nGenes[i] <- length(gene_in_region[[i]])
}
listlength <- max(my_region$nGenes)

#set a table to store gene name
genelist <- as.data.frame(matrix(nrow=length(gene_in_region),ncol=listlength))

for (row in 1:nrow(genelist)) {
  if (length(gene_in_region[[row]] != 0)){
    for (col in 1:length(gene_in_region[[row]])) {
    genelist[row,col] <- gene_in_region[[row]][col]
    }
  }
}

#save region and genes in csv
write.csv(cbind(my_region,genelist), 
          file = "region_with_gene.csv",
          row.names = F,
          na = "")
```

### Find gene function from genelist

```{r}
annotation_file <- "D:\\LEARN\\project\\R_workingdir\\MpTak_v6.1_func_annotation_1line.tsv"
annotation <- read.csv(annotation_file,header = F,sep = "\t")

for (i in 1:nrow(annotation)) {
  annotation$ID1[i] <- strsplit2(annotation$V1[i],split = "\\.")[1]
}
region_annotation <- subset(annotation,
                     annotation$ID1 %in% unique(genelist[genelist != "NA"]) )

annotation_file2 <- "D:\\LEARN\\project\\R_workingdir\\MpTak_v6.1_func_annotation.tsv"
annotation2 <- read.csv(annotation_file2,header = F,sep = "\t")

for (i in 1:nrow(annotation2)) {
  annotation2$ID1[i] <- strsplit2(annotation2$V1[i],split = "\\.")[1]
}



region_annotation2 <- subset(annotation2,
                     annotation2$ID1 %in% unique(genelist[genelist != "NA"]) )


#find my pi calculation/ nSNPs for these genes (from my_gff)
sub_pi <- subset(my_gff, my_gff$ID %in% region_annotation$ID1)
for (i in 1:nrow(region_annotation)) {
  for (j in 1:nrow(sub_pi)) {
    if (sub_pi$ID[j] == region_annotation$ID1[i]){
       region_annotation$pi[i] <- sub_pi$PI[j]
       region_annotation$nSNP[i] <- sub_pi$NUMBERofSNPs[j]
    }
    
  }
}

gene_table <- apply(region_annotation,2,as.character)
write.csv(gene_table, file = "Gene_annotationonleline_in_region.csv",row.names = F)
write.csv(sub_pi, file = "Gene_gff_in_region.csv",row.names = F)
remove(gene_table)
```

### Use chr8 8918596:9558653 as example

```{r}
region_chr8 <- subset(region_annotation2,
                     region_annotation2$ID1 %in% unique(genelist[16,]) )

region_chr8_oneline <- subset(region_annotation,
                     region_annotation$ID1 %in% unique(genelist[16,]) )


write.csv(region_chr8_oneline, file = "regionchr8genes.csv",row.names = F)
write.csv(region_chr8, file = "regionchr8genes-2.csv",row.names = F)
#average pi for each gene
mean(na.omit(region_chr8_oneline$pi))
```
